### PR TITLE
[Edge] header for nns-edge

### PR DESCRIPTION
**Pull request by jaeyun-jung**
initial commit to implement internal library for nnstreamer edge based on glib.

Signed-off-by: Jaeyun <jy1210.jung@samsung.com>

### CODE REVIEW

**Review by gichan-jang — APPROVED** [review]
LGTM!

File: gst/nnstreamer/tensor_query/nnstreamer_edge.h [hunk 1]
```diff
@@ -0,0 +1,165 @@
+/* SPDX-License-Identifier: LGPL-2.1-only */
```
Comment (myungjoo) [inline gst/nnstreamer/tensor_query/nnstreamer_edge.h:1]: It's good now. but when you've migrated to nnstreamer-edge.git or aitt.git, you may use Apache-2.0

File: gst/nnstreamer/tensor_query/nnstreamer_edge.h [hunk 2]
```diff
@@ -0,0 +1,165 @@
...
+ */
+typedef enum
+{
+  NNS_EDGE_ERROR_NONE = 0,
+  NNS_EDGE_ERROR_UNKNOWN = -1,
+  NNS_EDGE_ERROR_INVALID_PARAMETER = -2,
+  NNS_EDGE_ERROR_OUT_OF_MEMORY = -3,
+  NNS_EDGE_ERROR_CONNECTION_FAILURE = -4,
+  NNS_EDGE_ERROR_IO = -5,
+} nns_edge_error_e;
```
Comment (myungjoo) [inline gst/nnstreamer/tensor_query/nnstreamer_edge.h:47]: You may use Linux standard error code instead. (errno.h)

**Review by myungjoo — COMMENTED** [review]
LGTM in overall. PTAL: @anyj0527 @again4you

**Review by anyj0527 — APPROVED** [review]
LGTM ✔

File: gst/nnstreamer/tensor_query/nnstreamer_edge.h (hunk 2 above)
Comment (myungjoo) [reply gst/nnstreamer/tensor_query/nnstreamer_edge.h:48]: Having non Linux standard error code will cause some problems in the future. Please reuse Linux error code unless you have some specific reason not to.

**Review by myungjoo — COMMENTED** [review]
@again4you : Please review the exported header and merge if you think it's good enough. (nnstreamer_edge.h's API format)

File: gst/nnstreamer/tensor_query/nnstreamer_edge.h [hunk 3]
```diff
@@ -0,0 +1,166 @@
...
+/**
+ * @brief Callback called when nnstreamer-edge data is released.
+ */
+typedef void (*nns_edge_data_destroy_cb) (void *data);
+
+/**
+ * @brief Get registered handle. If not registered, create new handle and register it.
+ */
+int
+nns_edge_get_handle (const char *id, const char *topic, nns_edge_h *edge_h);
```
Comment (again4you) [inline gst/nnstreamer/tensor_query/nnstreamer_edge.h:89]: The name of Edge APIs is intuitive so easy to make a code. However, there is no example to refer to.
What about adding some pseudo-code for API usage? In my opinion, simple code without error checking is good enough as below.

```cpp
nns_edge_h handle;
nns_edge_data_h data_h;

/* create the edge handle and connect the MQTT broker */
nns_edge_get_handle (USER_ID, "object_detect/mobile_v3", &handle);
nns_edge_connect (handle, NNS_EDGE_PROTOCOL_MQTT, BROKER_IP, BROKER_PORT);

/* create the edge data to publish */
nns_edge_data_create (NNS_EDGE_DATA_TYPE_RAW_DATA, &data_h);
nns_edge_data_add (data_h, buffer, buffer_size, NULL);

/* publish single MQTT message */
nns_edge_publish (handle, data_h);

/* finalize handles */
nns_edge_data_destroy (data_h);
nns_edge_disconnect (handle);
nns_edge_release_handle (handle);
```

File: gst/nnstreamer/tensor_query/nnstreamer_edge_common.c [hunk 4]
```diff
@@ -0,0 +1,118 @@
...
+  unsigned int i;
+
+  ed = (nns_edge_data_s *) data_h;
+
+  if (!nns_edge_data_is_valid (data_h)) {
+    nns_edge_loge ("Invalid param, given edge data is invalid.");
+    return NNS_EDGE_ERROR_INVALID_PARAMETER;
+  }
+
+  ed->magic = 0xdeaddead;
```
Comment (again4you) [inline gst/nnstreamer/tensor_query/nnstreamer_edge_common.c:79]: Need to define the macro in `nnstreamer_edge_common.h`?

**Review by again4you — COMMENTED** [review]
The name of New Edge API is intuitive and the usage pattern is similar to ML API.  I think our customers can make the code easily. 👍 💯 

I added two opinions about this PR: the error code and the usage example. Others are not major issues so you can ignore them if you do not want.

File: gst/nnstreamer/tensor_query/nnstreamer_edge.h (hunk 2 above)
Comment (again4you) [reply gst/nnstreamer/tensor_query/nnstreamer_edge.h:48]: I added some comment for this issue.
* https://github.com/nnstreamer/nnstreamer/pull/3710#discussion_r865488072

File: gst/nnstreamer/tensor_query/nnstreamer_edge_common.c [hunk 5]
```diff
@@ -0,0 +1,118 @@
...
+    return NNS_EDGE_ERROR_INVALID_PARAMETER;
+  }
+
+  ed = g_try_new (nns_edge_data_s, 1);
+  if (!ed) {
+    nns_edge_loge ("Failed to allocate memory for edge data.");
+    return NNS_EDGE_ERROR_OUT_OF_MEMORY;
+  }
+
+  memset (ed, 0, sizeof (nns_edge_data_s));
```
[the remark this answers is not in the document]
Comment (jaeyun-jung) [reply gst/nnstreamer/tensor_query/nnstreamer_edge_common.c:55]: It is better to use predefined macros for zero-initialization, however, we should remove glib dependency later and IMO zero-init here is not bad for future work.

[the remark this answers is not in the document]
Comment (again4you) [reply gst/nnstreamer/tensor_query/nnstreamer_edge_common.c:55]: Ah! You are right. If then, primitive stdlib.h header functions are better. :+1:

File: gst/nnstreamer/tensor_query/nnstreamer_edge.h (hunk 1 above)
Comment (jaeyun-jung) [reply gst/nnstreamer/tensor_query/nnstreamer_edge.h:1]: marked todo (update license)

File: gst/nnstreamer/tensor_query/nnstreamer_edge_aitt.c [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+#include "config.h"
+#endif
+
+#include "nnstreamer_edge.h"
+#include "aitt_c.h"
+#include <stdio.h>
+#include <stdlib.h>
+
+#define DEBUG	0
+#define DEFAULT_QOS 0
```
[the remark this answers is not in the document]
Comment (jaeyun-jung) [reply gst/nnstreamer/tensor_query/nnstreamer_edge_aitt.c:23]: removed this macro. now aitt-implementation is todo item.

File: gst/nnstreamer/tensor_query/nnstreamer_edge.h [hunk 7]
```diff
@@ -0,0 +1,166 @@
...
+ */
+typedef enum
+{
+  NNS_EDGE_ERROR_NONE = 0,
+  NNS_EDGE_ERROR_UNKNOWN = -1,
+  NNS_EDGE_ERROR_INVALID_PARAMETER = -2,
+  NNS_EDGE_ERROR_OUT_OF_MEMORY = -3,
+  NNS_EDGE_ERROR_CONNECTION_FAILURE = -4,
+  NNS_EDGE_ERROR_IO = -5,
+} nns_edge_error_e;
```
[the remark this answers is not in the document]
Comment (jaeyun-jung) [reply gst/nnstreamer/tensor_query/nnstreamer_edge.h:48]: I updated draft, and marked todo to set detailed error (linux std error code). Please see the update.

File: gst/nnstreamer/tensor_query/nnstreamer_edge_aitt.c [hunk 8]
```diff
@@ -0,0 +1,46 @@
...
+typedef void *nns_edge_aitt_h;
+typedef void *nns_edge_aitt_msg_h;
+typedef void *nns_edge_aitt_sub_h;
+
+/**
+ * @brief Data structure for aitt handle.
+ */
+typedef struct
+{
+
```
[the remark this answers is not in the document]
Comment (jaeyun-jung) [reply gst/nnstreamer/tensor_query/nnstreamer_edge_aitt.c:38]: updated

File: gst/nnstreamer/tensor_query/nnstreamer_edge_common.h [hunk 9]
```diff
@@ -0,0 +1,74 @@
...
+  unsigned int num;
+  nns_edge_raw_data_s data[NNS_EDGE_DATA_LIMIT];
+} nns_edge_data_s;
+
+/** @todo add log util */
+#define nns_edge_logi g_info
+#define nns_edge_logw g_warning
+#define nns_edge_loge g_critical
+#define nns_edge_logd g_debug
+#define nns_edge_logf g_error
```
[the remark this answers is not in the document]
Comment (jaeyun-jung) [reply gst/nnstreamer/tensor_query/nnstreamer_edge_common.h:63]: marked todo for log utils in nnstreamer-edge.

File: gst/nnstreamer/tensor_query/nnstreamer_edge_common.h [hunk 10]
```diff
@@ -0,0 +1,74 @@
...
+ * @brief g_memdup() function replaced by g_memdup2() in glib version >= 2.68
+ */
+#if GLIB_USE_G_MEMDUP2
+#define _g_memdup g_memdup2
+#else
+#define _g_memdup g_memdup
+#endif
+
+#define NNS_EDGE_MAGIC 0xfeedfeed
+#define NNS_EDGE_MAGIC_IS_VALID(h) ((h) && (h)->magic == NNS_EDGE_MAGIC)
```
[the remark this answers is not in the document]
Comment (jaeyun-jung) [reply gst/nnstreamer/tensor_query/nnstreamer_edge_common.h:41]: updated

File: gst/nnstreamer/tensor_query/nnstreamer_edge_internal.c [hunk 11]
```diff
@@ -0,0 +1,330 @@
...
+{
+  nns_edge_handle_s *eh;
+
+  eh = (nns_edge_handle_s *) edge_h;
+
+  if (!NNS_EDGE_MAGIC_IS_VALID (eh))
+    return false;
+
+  return true;
+}
```
[the remark this answers is not in the document]
Comment (jaeyun-jung) [reply gst/nnstreamer/tensor_query/nnstreamer_edge_internal.c:44]: I updated declaration in common header.

File: gst/nnstreamer/tensor_query/nnstreamer_edge.h (hunk 3 above)
Comment (jaeyun-jung) [reply gst/nnstreamer/tensor_query/nnstreamer_edge.h:89]: marked todo (add sample code)

File: gst/nnstreamer/tensor_query/nnstreamer_edge_common.c (hunk 4 above)
Comment (jaeyun-jung) [reply gst/nnstreamer/tensor_query/nnstreamer_edge_common.c:79]: updated

File: gst/nnstreamer/tensor_query/nnstreamer_edge_common.h (hunk 10 above)
[the remark this answers is not in the document]
Comment (again4you) [reply gst/nnstreamer/tensor_query/nnstreamer_edge_common.h]: Verified!

File: gst/nnstreamer/tensor_query/nnstreamer_edge_internal.c (hunk 11 above)
[the remark this answers is not in the document]
Comment (again4you) [reply gst/nnstreamer/tensor_query/nnstreamer_edge_internal.c]: Verified!

File: gst/nnstreamer/tensor_query/nnstreamer_edge_common.c (hunk 4 above)
Comment (again4you) [reply gst/nnstreamer/tensor_query/nnstreamer_edge_common.c]: Verified!

File: gst/nnstreamer/tensor_query/nnstreamer_edge.h (hunk 3 above)
Comment (again4you) [reply gst/nnstreamer/tensor_query/nnstreamer_edge.h:90]: Verified!

**Review by again4you — APPROVED** [review]
The updated version looks good to me! :100:

**Pull request merged**